### PR TITLE
More partitions for resample_spatial with higher resolution.

### PR DESCRIPTION
**Pull request by EmileSonneveld**
Increase amount of partitions in proportion to the amount of resolution increased.

### CODE REVIEW

File: openeogeotrellis/geopysparkdatacube.py [hunk 1]
```diff
@@ -1222,15 +1222,31 @@ def resample_spatial(
...
-                max_level = max_level.repartition(spatialPartitions*10)
+            resolution_factor = (
+                cellsize_before[0] / resolution[0],
+                cellsize_before[1] / resolution[1]
+            )
+
+            # If the layer had a smaller tileSize specified, there will probably be enough partitions and this logic won't have effect
+            # If the user has a larger tileSize specified, it will probably be max 1024, then we would make more partitions
+            # Anyway, specifying tileSize is an undocumented feature, so we can assume default tileSize
+            default_tile_size = 256
```
Comment (EmileSonneveld) [inline openeogeotrellis/geopysparkdatacube.py:1239]: @jdries Is the reasoning behind this value ok?

Comment (jdries) [reply openeogeotrellis/geopysparkdatacube.py:1239]: 256 is not necessarily the default, there's also a feature that adapts it dynamically, even if the user does not set it. 
But I'm not sure why we need to guess it, if we can actually get it from the metadata?
I'm wondering if we should include the number of bands, and then perhaps some guesswork on the temporal dimension, to get a guesstimate of memory per partition?

Comment (EmileSonneveld) [reply openeogeotrellis/geopysparkdatacube.py:1239]: In normal cases, `currentTileLayout.tileCols` could be used, but in `TestMultipleDates.test_resample_spatial` that would be `5`, causing a lot of partitions (slowing the test down x10).

I did not yet find a nice solution that works for normal process graphs and this test. 

For some reason, `layoutCols` is 2 in `test_resample_spatial`, so checking on this value to be 1 to resort to a default won't work

Comment (EmileSonneveld) [reply openeogeotrellis/geopysparkdatacube.py]: I could use tileCols if I just updated the layer in test_resample_spatial to have more realistic proportions. The test suite runs smooth now, as well as the ERA5 resample test.

File: openeogeotrellis/geopysparkdatacube.py [hunk 2]
```diff
@@ -1242,13 +1254,28 @@ def resample_spatial(
...
         extent = max_level.layer_metadata.layout_definition.extent
         currentTileLayout: gps.TileLayout = max_level.layer_metadata.tile_layout
+        return (
+            (extent.xmax - extent.xmin) / (currentTileLayout.tileCols * currentTileLayout.layoutCols),
+            (extent.ymax - extent.ymin) / (currentTileLayout.tileRows * currentTileLayout.layoutRows)
+        )
 
-        currentResolutionX = (extent.xmax - extent.xmin) / (currentTileLayout.tileCols * currentTileLayout.layoutCols)
-        return currentResolutionX
+    def estimate_layer_size_in_pixels(self):
```
Comment (jdries) [inline openeogeotrellis/geopysparkdatacube.py:1266]: what about taking number of bands into account as well?

File: tests/test_multiple_date.py [hunk 3]
```diff
@@ -300,6 +319,47 @@ def elem(timestamp, value):
 
         return TiledRasterLayer.from_numpy_rdd(LayerType.SPACETIME, rdd, metadata)
 
+    def get_test_layer(self):
```
Comment (jdries) [inline tests/test_multiple_date.py:322]: note that there's also some test fixtures somewhere, perhaps those could be improved & reused, or this method could be converted into such a fixture?

File: openeogeotrellis/geopysparkdatacube.py [hunk 4]
```diff
@@ -1222,15 +1222,27 @@ def resample_spatial(
...
-            if(max_level.getNumPartitions() <= spatialPartitions and max_level.layer_type == gps.LayerType.SPACETIME and resolution < 0.5*self.get_cellsize_x()):
-                logging.info(f"Repartitioning datacube with {max_level.getNumPartitions()} partitions to {spatialPartitions*10} before resample_spatial.")
-                max_level = max_level.repartition(spatialPartitions*10)
+            resolution_factor = (
+                cellsize_before[0] / resolution[0],
+                cellsize_before[1] / resolution[1]
+            )
+
+            proposed_partition_count_tup = (
+                estimated_size_in_pixels_tup[0] * resolution_factor[0] / currentTileLayout.tileCols,
```
Comment (jdries) [inline openeogeotrellis/geopysparkdatacube.py:1237]: I'm wondering if this formula is actually a more complex way to write something like:
number_of_spatial_tiles*resolution_factor^2
but anyway, perhaps doesn't matter as I would go for a memory based estimate in the final solution, that we perhaps do after merging this PR?

**Review by jdries — CHANGES_REQUESTED** [review]

File: openeogeotrellis/geopysparkdatacube.py (hunk 4 above)
Comment (EmileSonneveld) [reply openeogeotrellis/geopysparkdatacube.py:1237]: It makes a slight difference when layer is not divisible by the tilesize. But is maybe not worth the complexity
In my SPI test it goes from `922` to `1128` when multiplying tile number directly.

File: tests/test_multiple_date.py (hunk 3 above)
Comment (EmileSonneveld) [reply tests/test_multiple_date.py]: True, moved!

File: openeogeotrellis/geopysparkdatacube.py (hunk 2 above)
Comment (EmileSonneveld) [reply openeogeotrellis/geopysparkdatacube.py]: Should not needed here. I added a comment `The repartitioning only considers the resolution change. It considers that the partitioning took already into account band count and the pixel type, sparse/not, ...`

**Review by jdries — APPROVED** [review]
Thanks and approved!
Lets merge this one, as it's probably better than the previous guesstimate.
Next step is computing memory per partition in a generic manner.

**Pull request merged**